libraries/persistent_store/src/storage.rs

```diff
@@ -34,6 +34,15 @@ pub enum StorageError {
 
     /// Implementation-specific error.
     CustomError,
+
+    /// libtock-rs Error
+    LibtockRsError(libtock_platform::ErrorCode),
```

Comment:
Yeah, it can be probably removed. I added it for debugging purposes in my process of getting a working MVP to get a more specific error message (i.e. in my case it was some userspace error and in the end `BadRVal`). With simply `CustomError` I could not locate the error.

Reply:
> Why do we need to add this? This should fit in the `CustomError` case. However, I can see why the `From` impl can only be defined here. So maybe it would make sense to provide a blanket impl:
> 
> ```rust
> impl From<T> for StorageError {
>     fn from(_: T) -> Self { Self::CustomError }
> }
> ```
This won't work as it conflicts with the core implementation:
```
error[E0119]: conflicting implementations of trait `std::convert::From<storage::StorageError>` for type `storage::StorageError`
  --> libraries/persistent_store/src/storage.rs:46:1
   |
46 | impl<T> From<T> for StorageError {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: conflicting implementation in crate `core`:
           - impl<T> From<T> for T;

```

Reply:
I see, then I see at least 2 options (which would be independent PRs):
1. Make `StorageError` generic over the `CustomError` and an associated type of `Storage`.
2. Add a `StorageError::custom(error: impl Debug) -> Self` function which returns `CustomError` by default (ignoring the error), and panics with the error when `cfg(feature = "debug_custom")`. Implementers of `Storage` would use `custom` to wrap their own error.

But if you don't need this debugging feature anymore, let's just keep the code like it was and remove those debugging helpers.

Reply:
Ok. Will bloat up the error handling in the app then because it can no longer automatically convert the error though.

Reply:
> But if you don't need this debugging feature anymore, let's just keep the code like it was and remove those debugging helpers.

I still think there should be a way to, at least optionally, have some context to the error. Should it be possible to implement the `Error` [trait](https://doc.rust-lang.org/stable/core/error/trait.Error.html)? This should make it easier for converting errors, also.

Reply:
> Should it be possible to implement the Error trait?

Interesting, I wasn't aware Error made it to core. Seems pretty new: https://github.com/rust-lang/rust/issues/103765. We would need to see if we use a recent enough compiler and if it doesn't bloat the binary size too much.

> Will bloat up the error handling in the app then because it can no longer automatically convert the error though.

This is true for solution (2) and status quo. For solution (1) I would expect an `impl<T> From<T> for StorageError<T>` to work. For solution (2) and status quo the bloat is anyway already there. Existing code already produces `CustomError` manually. This should not happen very often because it should only happen in the `impl Storage` methods, so it should be acceptable. (Also, it's only source bloat, not binary bloat.)